pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+	const TYPE_ID: [u8; 4] = *b"pool";
+}
+
+/// The result of epoch execution of a given tranch within a pool
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, Default)]
+pub struct EpochDetails<BalanceRatio> {
+	pub supply_fulfillment: Perquintill,
+	pub redeem_fulfillment: Perquintill,
+	pub token_price: BalanceRatio,
+}
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, Default)]
+pub struct EpochExecutionTranche<Balance, BalanceRatio> {
+	value: Balance,
+	price: BalanceRatio,
+	supply: Balance,
+	redeem: Balance,
+}
+
+/// The information for a currently executing epoch
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, Default)]
+pub struct EpochExecutionInfo<Balance, BalanceRatio> {
+	nav: Balance,
+	reserve: Balance,
+	tranches: Vec<EpochExecutionTranche<Balance, BalanceRatio>>,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use sp_std::convert::TryInto;
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type Balance: Member
+			+ Parameter
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaxEncodedLen
+			+ FixedPointOperand
+			+ From<u64>
+			+ TryInto<u64>;
+
+		/// A fixed-point number which represents the value of
+		/// one currency type in terms of another.
+		type BalanceRatio: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ FixedPointNumber<Inner = Self::Balance>;
+		type PoolId: Member + Parameter + Default + Copy + HasCompact + MaxEncodedLen;
+		type TrancheId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ HasCompact
+			+ MaxEncodedLen
+			+ Into<usize>
+			+ TryFrom<usize>;
+		type EpochId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ HasCompact
+			+ MaxEncodedLen
+			+ Zero
+			+ One
+			+ AddAssign;
+		type CurrencyId: Parameter + Copy;
+		type Tokens: MultiCurrency<
+			Self::AccountId,
+			Balance = Self::Balance,
+			CurrencyId = Self::CurrencyId,
+		>;
+
+		type LoanAmount: Into<Self::Balance>;
+		type NAV: PoolNAV<Self::PoolId, Self::LoanAmount>;
+
+		/// A conversion from a tranche ID to a CurrencyId
+		type TrancheToken: TrancheToken<Self>;
+		type Time: UnixTime;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn pool)]
+	pub type Pool<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::PoolId,
+		PoolDetails<T::AccountId, T::CurrencyId, T::EpochId, T::Balance>,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn order)]
+	pub type Order<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		TrancheLocator<T::PoolId, T::TrancheId>,
+		Blake2_128Concat,
+		T::AccountId,
+		UserOrder<T::Balance, T::EpochId>,
+		ValueQuery,
+	>;
```

Comment:
Is the storage used somewhere else than in the `supply` / `redeem` `Order::<T>::try_mutate()` calls? What is this used for? Are we using it for the solver?

Reply:
It will also be used for disbursement (walking through all epochs since the user submitted their order and transferring tokens into their account as needed based on how complete the epoch was)